server/src/main/java/org/opensearch/wlm/QueryGroupService.java

```diff
@@ -0,0 +1,146 @@
+import java.util.EnumMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Main service which will run periodically to track and cancel resource constraint violating tasks in QueryGroups
+ */
+public class QueryGroupService extends AbstractLifecycleComponent {
+    private static final Logger logger = LogManager.getLogger(QueryGroupService.class);
+
+    private final QueryGroupResourceUsageTrackerService queryGroupUsageTracker;
+    private final TimeValue queryGroupServiceRunIntervalInMillis;
+    private volatile Scheduler.Cancellable scheduledFuture;
+    private final ThreadPool threadPool;
+    private final ClusterService clusterService;
+    private final ClusterSettings clusterSettings;
+    private final Settings settings;
+    private Set<QueryGroup> activeQueryGroups = new HashSet<>();
+    private Set<QueryGroup> deletedQueryGroups = new HashSet<>();
+
+    /**
+     * Guice managed constructor
+     *
+     * @param queryGroupUsageTracker tracker service
+     * @param threadPool threadPool this will be used to schedule the service
+     */
+    @Inject
+    public QueryGroupService(
+        QueryGroupResourceUsageTrackerService queryGroupUsageTracker,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ClusterSettings clusterSettings,
+        Settings settings
+    ) {
+        this.queryGroupUsageTracker = queryGroupUsageTracker;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.clusterSettings = clusterSettings;
+        this.settings = settings;
+        this.queryGroupServiceRunIntervalInMillis = new TimeValue(
+            WorkloadManagementSettings.QUERYGROUP_SERVICE_RUN_INTERVAL_SETTING.get(settings)
+        );
+    }
+
+    /**
+     * run at regular interval
+     */
+    protected void doRun() {
+        Map<String, QueryGroupLevelResourceUsageView> queryGroupLevelResourceUsageViews = queryGroupUsageTracker
+            .constructQueryGroupLevelUsageViews();
+        this.activeQueryGroups = getActiveQueryGroups();
+        DefaultTaskCancellation defaultTaskCancellation = new DefaultTaskCancellation(
+            new DefaultTaskSelectionStrategy(),
+            queryGroupLevelResourceUsageViews,
+            activeQueryGroups,
+            deletedQueryGroups,
+            getNodeDuressTracker()::isNodeInDuress
+        );
+        defaultTaskCancellation.cancelTasks();
+    }
+
+    /**
+     * {@link AbstractLifecycleComponent} lifecycle method
+     */
+    @Override
+    protected void doStart() {
+        scheduledFuture = threadPool.scheduleWithFixedDelay(() -> {
+            try {
+                doRun();
+            } catch (Exception e) {
+                logger.debug("Exception occurred in Query Sandbox service", e);
+            }
+        }, this.queryGroupServiceRunIntervalInMillis, ThreadPool.Names.GENERIC);
+        queryGroupUsageTracker.constructQueryGroupLevelUsageViews();
```

Comment:
I think this is unnecessary since it is present in `doRun` method.

Reply:
good catch